File: packages/island-scene/preview-kit/serve.py
#!/usr/bin/env python3
"""Engage Island - local preview server (alternative to serve.mjs).

Serves the built game in ./game with Python 3's standard library only.

    python3 serve.py            this Mac only (127.0.0.1)
    python3 serve.py --lan      also other devices on the same Wi-Fi
    python3 serve.py --port 9000

Stop it with Control-C in the Terminal window it is running in.
"""
import argparse
import functools
import http.server
import os
import socket
import sys

ROOT = os.path.join(os.path.dirname(os.path.abspath(__file__)), "game")


class Handler(http.server.SimpleHTTPRequestHandler):
    extensions_map = {
        **http.server.SimpleHTTPRequestHandler.extensions_map,
        ".js": "text/javascript",
        ".mjs": "text/javascript",
        ".webp": "image/webp",
        ".wav": "audio/wav",
        ".json": "application/json",
    }

    def end_headers(self):
        self.send_header("Cache-Control", "no-cache")
        super().end_headers()

    def log_message(self, *args):
        pass  # quiet


def lan_addresses():
    """Best-effort: the address this Mac uses on the local network."""
    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        s.connect(("10.255.255.255", 1))  # no packet is sent for UDP connect
        addr = s.getsockname()[0]
        s.close()
        return [] if addr.startswith("127.") else [addr]
    except OSError:
        return []


def main():
    p = argparse.ArgumentParser()
    p.add_argument("--lan", action="store_true")
    p.add_argument("--port", type=int, default=8089)
    a = p.parse_args()
    if not os.path.isfile(os.path.join(ROOT, "index.html")):
        sys.exit(f"Can't find the game files next to this script (expected {ROOT}).")
    host = "0.0.0.0" if a.lan else "127.0.0.1"
    handler = functools.partial(Handler, directory=ROOT)
    try:
        httpd = http.server.ThreadingHTTPServer((host, a.port), handler)
    except OSError:
        sys.exit(f"Port {a.port} is already in use. Try: python3 serve.py --port {a.port + 1}")
    print("\n  Engage Island preview is running.\n")
    print(f"  On this Mac, open:   http://127.0.0.1:{a.port}/start.html")
    if a.lan:
        addrs = lan_addresses()
        print("")
        if addrs:
            print("  On an iPad or phone on the SAME Wi-Fi, open:")
            for ip in addrs:
                print(f"      http://{ip}:{a.port}/start.html")
        else:
            print("  No Wi-Fi/LAN address found on this Mac.")
        print("  (Anyone on this Wi-Fi can open it while this window is running.)")
    print("\n  To stop: press Control-C in this window.\n")
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        print("\n  Stopped.")


if __name__ == "__main__":
    main()

File: packages/island-scene/preview-kit/serve.mjs
#!/usr/bin/env node
/**
 * Engage Island — local preview server. Serves the built game in ./game.
 * Needs Node.js 18 or newer; no packages, nothing downloaded or installed.
 *
 *   node serve.mjs            this Mac only (127.0.0.1)
 *   node serve.mjs --lan      also other devices on the same Wi-Fi
 *   node serve.mjs --port 9000
 *
 * Stop it with Control-C in the Terminal window it is running in.
 */
import fs from "node:fs";
import http from "node:http";
import os from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = path.join(path.dirname(fileURLToPath(import.meta.url)), "game");
const args = process.argv.slice(2);
const lan = args.includes("--lan");
const portArg = args.indexOf("--port");
const port = portArg >= 0 ? Number(args[portArg + 1]) : 8089;
const host = lan ? "0.0.0.0" : "127.0.0.1";

const TYPES = {
  ".html": "text/html; charset=utf-8",
  ".js": "text/javascript; charset=utf-8",
  ".mjs": "text/javascript; charset=utf-8",
  ".css": "text/css; charset=utf-8",
  ".json": "application/json; charset=utf-8",
  ".webp": "image/webp",
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".svg": "image/svg+xml",
  ".wav": "audio/wav",
  ".mp3": "audio/mpeg",
  ".ogg": "audio/ogg",
  ".txt": "text/plain; charset=utf-8",
};

if (!fs.existsSync(path.join(ROOT, "index.html"))) {
  console.error(`Can't find the game files next to this script (expected ${ROOT}).`);
  process.exit(1);
}

const server = http.createServer((req, res) => {
  let rel;
  try {
    rel = decodeURIComponent(new URL(req.url ?? "/", "http://x").pathname);
  } catch {
    res.writeHead(400).end("Bad request");
    return;
  }
  let file = path.resolve(ROOT, "." + rel);
  if (file !== ROOT && !file.startsWith(ROOT + path.sep)) {
    res.writeHead(403).end("Forbidden");
    return;
  }
  if (fs.existsSync(file) && fs.statSync(file).isDirectory()) file = path.join(file, "index.html");
  fs.stat(file, (err, st) => {
    if (err || !st.isFile()) {
      res.writeHead(404, { "Content-Type": "text/plain; charset=utf-8" }).end("Not found");
      return;
    }
    res.writeHead(200, {
      "Content-Type": TYPES[path.extname(file).toLowerCase()] ?? "application/octet-stream",
      "Content-Length": st.size,
      "Cache-Control": "no-cache",
    });
    if (req.method === "HEAD") return res.end();
    fs.createReadStream(file).pipe(res);
  });
});

server.on("error", (e) => {
  if (e.code === "EADDRINUSE")
    console.error(`Port ${port} is already in use. Try: node serve.mjs --port ${port + 1}`);
  else console.error(e.message);
  process.exit(1);
});

server.listen(port, host, () => {
  console.log("");
  console.log("  Engage Island preview is running.");
  console.log("");
  console.log(`  On this Mac, open:   http://127.0.0.1:${port}/start.html`);
  if (lan) {
    const addrs = Object.values(os.networkInterfaces())
      .flat()
      .filter((a) => a && a.family === "IPv4" && !a.internal)
      .map((a) => a.address);
    console.log("");
    if (addrs.length) {
      console.log("  On an iPad or phone on the SAME Wi-Fi, open:");
      for (const a of addrs) console.log(`      http://${a}:${port}/start.html`);
    } else console.log("  No Wi-Fi/LAN address found on this Mac.");
    console.log("  (Anyone on this Wi-Fi can open it while this window is running.)");
  }
  console.log("");
  console.log("  To stop: press Control-C in this window.");
  console.log("");
});
